adapters/tlsio_appleios.c

```diff
@@ -0,0 +1,508 @@
+    return (CONCRETE_IO_HANDLE)tlsio_instance;
+}
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_021: [ The tlsio_appleios_destroy shall destroy a created instance of the tlsio for iOS identified by the CONCRETE_IO_HANDLE. ]*/
+void tlsio_appleios_destroy(CONCRETE_IO_HANDLE tlsio_handle)
+{
+    TLS_IO_INSTANCE* tlsio_instance = (TLS_IO_INSTANCE*)tlsio_handle;
+
+    if (tlsio_handle == NULL)
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_024: [ If the tlsio_handle is NULL, the tlsio_appleios_destroy shall not do anything. ]*/
+        LogError("NULL TLS handle.");
+    }
+    else
+    {
+        if ((tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPENING) ||
+            (tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPEN) ||
+            (tlsio_instance->state == TLSIO_APPLEIOS_STATE_CLOSING))
+        {
+            /* Codes_SRS_TLSIO_APPLEIOS_32_026: [ If the tlsio state is TLSIO_APPLEIOS_STATE_OPENING, TLSIO_APPLEIOS_STATE_OPEN, or TLSIO_APPLEIOS_STATE_CLOSING, the tlsio_appleios_destroy shall close destroy the tlsio, but log an error. ]*/
+            LogError("TLS destroyed with a SSL connection still active.");
+        }
+
+        /* Codes_SRS_TLSIO_APPLEIOS_32_022: [ The tlsio_appleios_destroy shall free the memory allocated for tlsio_instance. ]*/
+		CFRelease(tlsio_instance->remote_host);
+		CFRelease(tlsio_instance->sockRead);
+		CFRelease(tlsio_instance->sockWrite);
+        free(tlsio_instance);
+    }
+}
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_026: [ The tlsio_appleios_open shall star the process to open the ssl connection with the host provided in the tlsio_appleios_create. ]*/
+int tlsio_appleios_open_async(
+    CONCRETE_IO_HANDLE tlsio_handle,
+    ON_IO_OPEN_COMPLETE on_io_open_complete,
+    void* on_io_open_complete_context,
+    ON_BYTES_RECEIVED on_bytes_received,
+    void* on_bytes_received_context,
+    ON_IO_ERROR on_io_error,
+    void* on_io_error_context)
+{
+    int result = 0;
+    TLS_IO_INSTANCE* tlsio_instance = (TLS_IO_INSTANCE*)tlsio_handle;
+
+    if (tlsio_handle == NULL)
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_036: [ If the tlsio_handle is NULL, the tlsio_appleios_open shall not do anything, and return _LINE_. ]*/
+        LogError("NULL TLS handle.");
+        result = __FAILURE__;
+    }
+    else
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_004: [ The tlsio_appleios shall call the callbacks functions defined in the `xio.h`. ]*/
+        /* Codes_SRS_TLSIO_APPLEIOS_32_006: [ The tlsio_appleios shall return the status of all async operations using the callbacks. ]*/
+        /* Codes_SRS_TLSIO_APPLEIOS_32_007: [ If the callback function is set as NULL. The tlsio_appleios shall not call anything. ]*/
+        /* Codes_SRS_TLSIO_APPLEIOS_32_028: [ The tlsio_appleios_open shall store the provided on_io_open_complete callback function address. ]*/
+        tlsio_instance->on_io_open_complete = on_io_open_complete;
```

Comment:
>on_io_open_complete [](start = 46, length = 19)

on_io_open_complete, on_bytes_received, and on_io_error are mandatories. Please test if they are not NULL.

Reply:
For more information, take a look at devdoc/tlsio_requirements.md

---
In reply to: [141107266](https://github.com/Azure/azure-c-shared-utility/pull/109#discussion_r141107266) [](ancestors = 141107266)